src/html.rs

```diff
@@ -32,1559 +46,1730 @@ pub fn format_document_with_plugins<'a>(
     }
 
-    fn render_raw<'a>(&mut self, node: &'a AstNode<'a>, entering: bool) -> io::Result<bool> {
-        let NodeValue::Raw(ref literal) = node.data.borrow().value else {
-            panic!()
-        };
+    Ok(RenderMode::HTML)
+}
 
-        // No sourcepos.
-        if entering {
-            self.output.write_all(literal.as_bytes())?;
+fn render_superscript<'a>(
+    context: &mut Context,
+    node: &'a AstNode<'a>,
+    entering: bool,
+) -> io::Result<RenderMode> {
+    // Unreliable sourcepos.
+    if entering {
+        context.write_all(b"<sup")?;
+        if context.options.render.experimental_inline_sourcepos {
+            render_sourcepos(context, node)?;
         }
-
-        Ok(true)
+        context.write_all(b">")?;
+    } else {
+        context.write_all(b"</sup>")?;
     }
 
-    #[cfg(feature = "shortcodes")]
-    fn render_short_code<'a>(&mut self, node: &'a AstNode<'a>, entering: bool) -> io::Result<bool> {
-        let NodeValue::ShortCode(ref nsc) = node.data.borrow().value else {
-            panic!()
-        };
+    Ok(RenderMode::HTML)
+}
 
-        // Nowhere to put sourcepos.
-        if entering {
-            self.output.write_all(nsc.emoji.as_bytes())?;
+fn render_underline<'a>(
+    context: &mut Context,
+    node: &'a AstNode<'a>,
+    entering: bool,
+) -> io::Result<RenderMode> {
+    // Unreliable sourcepos.
+    if entering {
+        context.write_all(b"<u")?;
+        if context.options.render.experimental_inline_sourcepos {
+            render_sourcepos(context, node)?;
         }
+        context.write_all(b">")?;
+    } else {
+        context.write_all(b"</u>")?;
+    }
+
+    Ok(RenderMode::HTML)
+}
+
+fn render_wiki_link<'a>(
+    context: &mut Context,
+    node: &'a AstNode<'a>,
+    entering: bool,
+) -> io::Result<RenderMode> {
+    let NodeValue::WikiLink(ref nl) = node.data.borrow().value else {
+        panic!()
+    };
 
-        Ok(true)
+    // Unreliable sourcepos.
+    if entering {
+        context.write_all(b"<a")?;
+        if context.options.render.experimental_inline_sourcepos {
+            render_sourcepos(context, node)?;
+        }
+        context.write_all(b" href=\"")?;
+        let url = nl.url.as_bytes();
+        if context.options.render.unsafe_ || !dangerous_url(url) {
+            context.escape_href(url)?;
+        }
+        context.write_all(b"\" data-wikilink=\"true")?;
+        context.write_all(b"\">")?;
+    } else {
+        context.write_all(b"</a>")?;
     }
 
-    fn render_spoiler_text<'a>(
-        &mut self,
-        node: &'a AstNode<'a>,
-        entering: bool,
-    ) -> io::Result<bool> {
-        // Unreliable sourcepos.
-        if entering {
-            self.output.write_all(b"<span")?;
-            if self.options.render.experimental_inline_sourcepos {
-                self.render_sourcepos(node)?;
+    Ok(RenderMode::HTML)
+}
+
+// Helpers
+
+fn collect_text<'a>(node: &'a AstNode<'a>, output: &mut Vec<u8>) {
```

Comment:
We might consider making this public. In overriding `render_heading`, I need this function.